contracts/WooFeeManager.sol

```diff
@@ -59,87 +60,69 @@ contract WooFeeManager is InitializableOwnable, ReentrancyGuard, IWooFeeManager
+    function collectFee(uint256 amount, address brokerAddr) external override {
         TransferHelper.safeTransferFrom(quoteToken, msg.sender, address(this), amount);
-        (uint256 userRewardRate, uint256 brokerRewardRate) = rewardManager.getRewardInfo(broker);
-        rewardManager.addReward(tx.origin, amount.mulFloor(userRewardRate));
-        rewardManager.addReward(broker, amount.mulFloor(brokerRewardRate));
+
+        // Step 1: distribute rebate if needed
+        uint256 rebateRate = rebateManager.rebateRate(brokerAddr);
+        uint256 rebateAmount = amount.mulFloor(rebateRate);
+        if (rebateAmount > 0) {
+            TransferHelper.safeApprove(quoteToken, address(rebateManager), rebateAmount);
+            rebateManager.addRebate(brokerAddr, rebateAmount);
+        }
+        uint256 feeAfterRebate = amount.sub(rebateAmount);
+
+        // Step 2: distribute to vault treasury
+        uint256 vaultRewardAmount = feeAfterRebate.mul(vaultRewardPercent).div(100);
+        if (vaultRewardAmount > 0) {
+            TransferHelper.safeApprove(quoteToken, address(vaultManager), vaultRewardAmount);
+            vaultManager.addReward(vaultRewardAmount);
+        }
     }
 
     /* ----- Admin Functions ----- */
 
-    /// @dev Set fee rate.
-    function setFeeRate(address token, uint256 newFeeRate) external onlyOwner {
+    function setFeeRate(address token, uint256 newFeeRate) external override onlyOwner {
         require(newFeeRate <= 1e16, 'WooFeeManager: FEE_RATE>1%');
         feeRate[token] = newFeeRate;
         emit FeeRateUpdated(token, newFeeRate);
     }
 
-    /// @dev Set WooPP.
-    function setWooPP(address newWooPP) external onlyOwner {
-        wooPP = newWooPP;
-    }
-
-    /// @dev Set reward token.
-    function setRewardToken(address newRewardToken) external onlyOwner {
-        rewardToken = newRewardToken;
-    }
-
-    /// @dev Swap quote token to reward token.
-    /// @param amount the amount of quote token to swap
-    function swap(uint256 amount) external onlyOwner {
-        _swap(amount);
-    }
-
-    /// @dev Withdraw the token.
-    /// @param token the token to withdraw
-    /// @param to the destination address
-    /// @param amount the amount to withdraw
-    function withdraw(
-        address token,
-        address to,
-        uint256 amount
-    ) public nonReentrant onlyOwner {
+    function emergencyWithdraw(address token, address to) external onlyOwner {
         require(token != address(0), 'WooFeeManager: token_ZERO_ADDR');
         require(to != address(0), 'WooFeeManager: to_ZERO_ADDR');
+        uint256 amount = IERC20(token).balanceOf(address(this));
         TransferHelper.safeTransfer(token, to, amount);
         emit Withdraw(token, to, amount);
     }
 
-    function withdrawAll(address token, address to) external onlyOwner {
-        withdraw(token, to, IERC20(token).balanceOf(address(this)));
+    function setRebateManager(address newRebateManager) external onlyOwner {
```

Comment:
Why `rebateManager` and `vaultManager` are not set at beginning? This contract cannot work without them (maybe unless setting `rebateRate` and `vaultManager` to 0.

Reply:
Otherwise too many classes need to be initialized before WooPP; but need to set up WooPP for them later on.